src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +249,140 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
+            if (serverNamePartsByComma.Length < 2 && serverNamePartsByBackSlash.Length < 2)
             {
+                return new SNITCPHandle(fullServerName, DEFAULT_SQL_SVR_PORT, timerExpire, callbackObject, parallel);
+            }
+            else if (serverNamePartsByComma.Length == 2 && serverNamePartsByBackSlash.Length < 2)
+            {
+                string hostName = serverNamePartsByComma[0];
+                int portNumber = -1;
                 try
                 {
-                    portNumber = ushort.Parse(serverAndPortParts[1]);
+                    portNumber = ushort.Parse(serverNamePartsByComma[1]);
+                    return new SNITCPHandle(hostName, portNumber, timerExpire, callbackObject, parallel);
                 }
                 catch (Exception e)
                 {
                     SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
                     return null;
                 }
             }
-            else if (serverAndPortParts.Length > 2)
+            else if (serverNamePartsByComma.Length < 2 && serverNamePartsByBackSlash.Length == 2)
+            {
+                string hostName = serverNamePartsByBackSlash[0];
+                string instanceName = serverNamePartsByBackSlash[1];
+                int portNumber = -1;
+                try
+                {
+                    portNumber = GetPortByInstanceName(hostName, instanceName);
+                    return new SNITCPHandle(hostName, portNumber, timerExpire, callbackObject, parallel);
+                }
+                catch (Exception e)
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
+                    return null;
+                }
+            }
+            else
             {
                 SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                 return null;
             }
+        }
+
+        /// <summary>
+        /// Sends CLNT_UCAST_INST request for given instance name to SQL Sever Browser, and receive SVR_RESP from the Browser.
+        /// </summary>
+        /// <param name="browserEP">SQL Sever Browser IPEndPoint</param>
+        /// <param name="instanceName">instance name for CLNT_UCAST_INST request</param>
+        /// <returns>SVR_RESP packets from SQL Sever Browser</returns>
+        private static byte[] SendInstanceNameRequest(string browserHostname, string instanceName)
+        {
+            if (String.IsNullOrWhiteSpace(browserHostname) || String.IsNullOrWhiteSpace(instanceName))
+            {
+                throw new ArgumentException();
+            }
+
+            int byteCount = Encoding.ASCII.GetByteCount(instanceName);
+            byte[] requestPacket = new byte[byteCount + 1];
+            requestPacket[0] = 0x04;
```

Comment:
Nit: Sorry I didn't ask about this in the previous review, but what does `0x04` mean? Should this be put into a constant to give it a more descriptive name? e.g.:

```c#
const int DescriptiveNameForConstant = 0x04;
int byteCount = Encoding.ASCII.GetByteCount(instanceName);
byte[] requestPacket = new byte[byteCount + 1];
requestPacket[0] = DescriptiveNameForConstant;
```

where `DescriptiveNameForConstant` is replaced with the better name?